PyInstaller/building/build_main.py

```diff
@@ -389,71 +407,72 @@ def assemble(self):
+                # This module is retained in the cache, as a subsequently run
+                # post-graph hook could import this module as a hidden import.
+                if module_node is None:
                     continue
-                # Skip hook if not the right Node type.
-                node_type = type(from_node).__name__
-                if node_type not in VALID_MODULE_TYPES:
+
+                # If this module is unimportable, permanently ignore it.
+                if type(module_node).__name__ not in VALID_MODULE_TYPES:
+                    hooked_module_names.add(module_name)
                     continue
 
-                # Run all post-graph hooks for this module.
-                for hook_file in hooks_cache[imported_name]:
-                    # Import hook module from a file.
-                    imphook_object = ImportHook(imported_name, hook_file)
-                    # Expand module dependency graph.
-                    imphook_object.update_dependencies(self.graph)
-                    # Update cache of binaries and datas.
-                    additional_files_cache.add(imported_name, imphook_object.binaries, imphook_object.datas)
-
-                # Append applied hooks to the list 'applied_hooks'.
-                # These will be removed after the inner loop finishs.
-                # It also is a marker that iteration over hooks should
-                # continue.
-                applied_hooks.append(imported_name)
-
-            ### All hooks from cache were traversed - stop or run again.
-            if not applied_hooks:  # Empty list.
-                # No new hook was applied - END of hooks processing.
-                break
-            else:
-                # Remove applied hooks from the cache - its not
-                # necessary apply then again.
-                hooks_cache.remove(applied_hooks)
-                # Run again - reset list 'applied_hooks'.
-                applied_hooks = []
+                # For each hook script for this module...
+                for module_hook in module_hooks:
+                    # Run this script's post-graph hook if any.
+                    module_hook.post_graph()
 
+                    # Cache all external dependencies listed by this script
+                    # after running this hook, which could add dependencies.
+                    additional_files_cache.add(
+                        module_name,
+                        module_hook.scope.binaries,
+                        module_hook.scope.datas)
+
+                # Prevent this module's hooks from being run again.
+                hooked_module_names.add(module_name)
+
+            # If no post-graph hooks were run, terminate iteration.
+            if not hooked_module_names:
+                break
 
+            # Else, prevent run post-graph hooks from being run again.
+            module_hook_cache.uncache_modules(hooked_module_names)
```

Comment:
I'd move this in front of the `if not hooked_module_names: break` line. This makes it easier to understand. (Yes, I know, if `not hooked_module_names`, calling this function is a no-op.)

Or write it the other way round: `if hooked_module_names: ... uncache() else: break`.